File: scripts/gates/check_mcp_policy.py
#!/usr/bin/env python3
"""Zero-network gate for the default-deny MCP exposure and runtime contract."""

from __future__ import annotations

import ast
import json
from pathlib import Path
import subprocess
import sys


def root() -> Path:
    return Path(__file__).resolve().parents[2]


def run_selftest(repo: Path) -> int:
    tools_test = subprocess.run(
        [sys.executable, str(repo / "loopctl" / "mcp_tools.py"), "--selftest"],
        cwd=repo,
        check=False,
    )
    if tools_test.returncode != 0:
        return tools_test.returncode

    sys.path.insert(0, str(repo / "loopctl"))
    import mcp_runtime  # noqa: PLC0415

    red = 0

    def case(name: str, got, want) -> None:
        nonlocal red
        if got != want:
            print(f"SELFTEST case failed — {name}: got {got!r}, want {want!r}", file=sys.stderr)
            red = 1

    commit = mcp_runtime.git(repo, "rev-parse", "HEAD")
    tree = mcp_runtime.git(repo, "rev-parse", f"{commit}^{{tree}}")
    tools, modules, policy_digest = mcp_runtime.load_surface(repo, commit)
    case("bounded-policy-tool-count", 0 < len(tools) < 20, True)
    case("policy-digest-length", len(policy_digest), 64)
    case("every-tool-module-is-selected", all(tool["_policy"]["module"] in modules for tool in tools), True)
    try:
        mcp_runtime.resolve_ref(repo, "HEAD")
    except mcp_runtime.McpError:
        pass
    else:
        case("mutable-head-is-refused", "accepted", "McpError")
    try:
        mcp_runtime.to_argv(tools[0], {"force_receipt": "/tmp/escape"})
    except mcp_runtime.McpError:
        pass
    else:
        case("absolute-server-path-is-refused", "accepted", "McpError")

    response = mcp_runtime.handle(
        {"jsonrpc": "2.0", "id": 1, "method": "tools/list"},
        repo,
        commit,
        tree,
        tools,
        modules,
        policy_digest,
    )
    case("tools-list-hides-policy", all("_policy" not in tool for tool in response["result"]["tools"]), True)
    response = mcp_runtime.handle(
        {"jsonrpc": "2.0", "id": 2, "method": "tools/call", "params": {"name": "loopctl_macro_run"}},
        repo,
        commit,
        tree,
        tools,
        modules,
        policy_digest,
    )
    case("unexposed-tool-is-unknown", "error" in response, True)

    syntax = ast.parse((repo / "loopctl" / "mcp_runtime.py").read_text(encoding="utf-8"))
    symlink_calls = [
        node
        for node in ast.walk(syntax)
        if isinstance(node, ast.Call)
        and isinstance(node.func, ast.Attribute)
        and node.func.attr == "symlink_to"
    ]
    case("no-owner-dependency-symlink", len(symlink_calls), 0)

    parent: Path | None = None
    with mcp_runtime.disposable_worktree(repo, commit) as (_, worktree):
        parent = worktree.parent
        closure = mcp_runtime.module_closure(tools[0]["_policy"]["module"], modules)
        kept, removed = mcp_runtime.prune_worktree(
            worktree,
            mcp_runtime.closure_prefixes(closure, modules),
        )
        case("selected-closure-keeps-files", kept > 0, True)
        case("selected-closure-removes-files", removed > 0, True)
        case("unselected-notebooklm-is-absent", (worktree / "notebooklm").exists(), False)
    case("disposable-worktree-is-cleaned", parent is not None and not parent.exists(), True)

    print("SELFTEST " + ("GREEN" if not red else "RED"))
    return red


def check(repo: Path) -> int:
    process = subprocess.run(
        [
            sys.executable,
            str(repo / "loopctl" / "mcp_tools.py"),
            str(repo / "loopctl" / "contract.json"),
            "--policy",
            str(repo / ".arena" / "mcp-policy.json"),
        ],
        cwd=repo,
        text=True,
        capture_output=True,
        check=False,
    )
    if process.returncode != 0:
        sys.stderr.write(process.stderr)
        return process.returncode
    try:
        value = json.loads(process.stdout)
    except json.JSONDecodeError as exc:
        print(f"MCP policy FATAL: generator output is not JSON: {exc}", file=sys.stderr)
        return 64
    tools = value.get("tools")
    if not isinstance(tools, list) or not tools:
        print("MCP policy RED: explicit policy generated no tools", file=sys.stderr)
        return 2
    names = [tool.get("name") for tool in tools]
    if len(names) != len(set(names)):
        print("MCP policy RED: duplicate tool names", file=sys.stderr)
        return 2
    if any(any(key.startswith("_") for key in tool) for tool in tools):
        print("MCP policy RED: public generator leaked internal fields", file=sys.stderr)
        return 2
    forbidden = {
        "loopctl_macro_run",
        "loopctl_mcp_serve",
        "loopctl_container_build",
        "loopctl_openwiki_run",
        "loopctl_notebooklm_run",
        "loopctl_equivalence_run",
    }
    leaked = sorted(forbidden.intersection(names))
    if leaked:
        print(f"MCP policy RED: dangerous tools exposed: {leaked}", file=sys.stderr)
        return 2
    print(f"PASS default-deny MCP policy ({len(tools)} tools)")
    return 0


def main(argv: list[str]) -> int:
    repo = root()
    if argv == ["--selftest"]:
        return run_selftest(repo)
    if argv:
        print(__doc__.strip(), file=sys.stderr)
        return 64
    return check(repo)


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: scripts/gates/check_mcp_policy.ts
#!/usr/bin/env bun
import { existsSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import {
  McpError,
  buildTools,
  closurePrefixes,
  createWorkspace,
  gitText,
  loadSurface,
  moduleClosure,
  pruneWorktree,
  publicTool,
  resolveRef,
  toArgv,
  type LoopContract,
} from "../../loopctl/mcp_core.ts";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(HERE, "../..");

function check(): number {
  const commit = gitText(ROOT, ["rev-parse", "HEAD"]);
  const surface = loadSurface(ROOT, commit);
  if (!surface.tools.length || surface.tools.length >= 20) {
    throw new McpError(`explicit policy generated an invalid tool count: ${surface.tools.length}`);
  }
  const forbidden = new Set([
    "loopctl_macro_run",
    "loopctl_mcp_serve",
    "loopctl_container_build",
    "loopctl_openwiki_run",
    "loopctl_notebooklm_run",
    "loopctl_equivalence_run",
  ]);
  const leaked = surface.tools.filter((tool) => forbidden.has(tool.name));
  if (leaked.length) {
    throw new McpError(`dangerous tools exposed: ${leaked.map((tool) => tool.name)}`);
  }
  if (
    surface.tools.some((tool) =>
      Object.keys(publicTool(tool)).some((key) => key.startsWith("_")),
    )
  ) {
    throw new McpError("public tool projection leaked internal fields");
  }
  const runtime = readFileSync(resolve(ROOT, "loopctl/mcp_runtime.ts"), "utf8");
  if (/symlink(?:Sync|To)|node_modules.*owner/i.test(runtime)) {
    throw new McpError("Bun runtime contains owner dependency borrowing");
  }
  console.log(`PASS default-deny Bun MCP policy (${surface.tools.length} tools)`);
  return 0;
}

function selftest(): number {
  let red = 0;
  const expect = (name: string, condition: boolean): void => {
    if (condition) return;
    console.error(`SELFTEST case failed — ${name}`);
    red = 1;
  };

  const contract: LoopContract = {
    modes: { run: "run" },
    commands: [
      {
        loop: "demo",
        mode: "run",
        target: "x",
        required: [],
        optional: ["--json"],
      },
    ],
  };
  expect("no-policy-is-empty", buildTools(contract, null).length === 0);

  let mutableRefRejected = false;
  try {
    resolveRef(ROOT, "HEAD");
  } catch {
    mutableRefRejected = true;
  }
  expect("mutable-head-is-rejected", mutableRefRejected);

  const commit = gitText(ROOT, ["rev-parse", "HEAD"]);
  const surface = loadSurface(ROOT, commit);
  const nonCarrier = surface.tools.find((tool) => !tool._carrier);
  expect("policy-has-a-non-carrier-control", Boolean(nonCarrier));
  if (nonCarrier) {
    let hostPathRejected = false;
    try {
      toArgv(nonCarrier, { force_receipt: "/tmp/escape" });
    } catch {
      hostPathRejected = true;
    }
    expect("server-host-path-is-rejected", hostPathRejected);
  }

  const tool = surface.tools[0];
  expect("policy-has-an-exposed-tool", Boolean(tool));
  if (tool) {
    const workspace = createWorkspace(ROOT, commit);
    const parent = workspace.base;
    try {
      const closure = moduleClosure(tool._policy.module, surface.modules);
      const result = pruneWorktree(
        workspace.worktree,
        closurePrefixes(closure, surface.modules),
      );
      expect("pruning-keeps-and-removes-files", result.kept > 0 && result.removed > 0);
      if (tool._policy.module !== "notebooklm") {
        expect(
          "unselected-module-bytes-are-absent",
          !existsSync(resolve(workspace.worktree, "notebooklm/workflow.py")),
        );
      }
    } finally {
      workspace.cleanup();
    }
    expect("workspace-is-cleaned", !existsSync(parent));
  }

  console.log(`SELFTEST ${red ? "RED" : "GREEN"}`);
  return red;
}

function main(argv: string[]): number {
  try {
    if (argv[0] === "--selftest") return selftest();
    if (argv.length) {
      throw new McpError("no arguments are accepted except --selftest");
    }
    return check();
  } catch (error) {
    console.error(
      `MCP policy RED: ${String(error instanceof Error ? error.message : error)}`,
    );
    return error instanceof McpError ? 2 : 64;
  }
}

const invoked =
  process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (invoked) process.exit(main(process.argv.slice(2)));
